### PR TITLE
[BUG] Resolve bugs in the TapNet Tensorflow implementation

### LINKED ISSUE #9446: [BUG] TapNet TF: RP dimension uses series_length and CNN weights not shared (paper mismatch)
**Describe the bug**
There are 3 critical issues in the TapNet estimator implementation in Tensorflow. 
- In the convolutional layers, the tf implementation uses kernel_size[0] everywhere, but according to the [original TapNet paper](https://xuczhang.github.io/papers/aaai20_tapnet_full.pdf), its appropriate to use all 3 of the specified kernel sizes.
- The paper mentions parameter sharing for the convolutional layers, however the tensorflow implementation does not implement this.
- When computing the random projection dimensions, it uses the series length, but later selects channels along axis 2. The correct way is to handle the case for `n_dims <= 1` and use just `n_dims` (channels) to calculate the rp dimensions. 

> This issue was identified while working on #9322 

**Additional context**
Reviewer comment: https://github.com/sktime/sktime/pull/9412#issuecomment-3937094030
TapNet paper link: https://xuczhang.github.io/papers/aaai20_tapnet_full.pdf
TapNet PyTorch implementation PR (with necessary changes): https://github.com/sktime/sktime/pull/9412

**Versions**
<details>

System:
    python: 3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:17:27) [MSC v.1929 64 bit (AMD64)]
executable: C:\ProgramData\anaconda3\python.exe
   machine: Windows-11-10.0.26200-SP0

Python dependencies:
          pip: 25.3
       sktime: 0.40.1
      sklearn: 1.7.0
       skbase: 0.13.1
        numpy: 1.26.4
        scipy: 1.16.0
       pandas: 2.3.3
   matplotlib: 3.9.2
       joblib: 1.5.1
huggingface-hub: 0.24.6
        numba: 0.62.1
     pmdarima: None
pytorch-forecasting: None
   skforecast: None
        skpro: None
statsforecast: None
  statsmodels: 0.14.2
 transformers: 4.44.2
      tsfresh: None
      tslearn: None
        torch: 2.9.1
   tensorflow: None
</details>

### PR DESCRIPTION
**Pull request by srupat**
#### Reference Issues/PRs
Fixes #9446 

#### What does this implement/fix? Explain your changes.
Fixes the 3 minor bugs in the Tensorflow TapNet implementation.
- In the convolutional layers, the tf implementation uses kernel_size[0] everywhere, but according to the [original TapNet paper](https://xuczhang.github.io/papers/aaai20_tapnet_full.pdf), its appropriate to use all 3 of the specified kernel sizes.
- The paper mentions parameter sharing for the convolutional layers, however the tensorflow implementation does not implement this.
- When computing the random projection dimensions, it uses the series length, but later selects channels along axis 2. The correct way is to handle the case for n_dims <= 1 and use just n_dims (channels) to calculate the rp dimensions.

#### What should a reviewer concentrate their feedback on?
- The logic for setting `rp_dim` using `n_dims`. Currently used `rp_dim = max(1, floor(n_dims*2/3))` for compatibility, however, there are some considerations which I have addressed [here](https://github.com/sktime/sktime/pull/9412#discussion_r2835906752).
- Parameter sharing in the convolutional layers.
- Appropriate kernel_size use in the convolutional layers.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Comment by srupat** [comment]
@RecreationalMath could you please review this too?

File: sktime/networks/tapnet.py [hunk 1]
```diff
@@ -192,7 +193,33 @@ def build_network(self, input_shape, **kwargs):
             # Convolutional Network
             # input ts: # N * C * L
             if self.use_rp:
-                self.conv_1_models = keras.Sequential()
+                conv_1 = keras.layers.Conv1D(
```
Comment (RecreationalMath) [inline sktime/networks/tapnet.py]: This block here implements full sharing (all layers are shared) but if you look at the [original codebase](https://github.com/kdd2019-tapnet/tapnet/blob/master/models.py) (check `__init__` and `forward`) you'll find 1st layer is separate and remaining layers are shared (partial sharing). The sharing as per the github repo is:
- Layer 1 (conv_1): Independent per group, each group has its own first conv, allowing it to learn group specific features from its particular random channel subset
- Layers 2–3 (conv_2, conv_3) and all BN: Shared across groups

Also see section `4.3 Multivariate Time Series Encoding` in the original paper. It specifies the exact nature of parameter sharing.

So while the original `tapnet` in sktime had the bug of undersharing, it seems this PR's version overshares.

File: sktime/networks/tapnet.py [hunk 2]
```diff
@@ -171,8 +171,9 @@ def build_network(self, input_shape, **kwargs):
         input_layer = keras.layers.Input(input_shape)
 
         if self.rp_params[0] < 0:
-            dim = input_shape[0]
-            self.rp_params = [3, math.floor(dim * 2 / 3)]
+            dim = input_shape[1]
+            rp_dim = max(1, math.floor(dim * 2 / 3))
```
Comment (RecreationalMath) [inline sktime/networks/tapnet.py]: `rp_dim = max(1, 0) = 1` would essentially select 1 channel out of 1 and then concatenate all of them because each of the `rp_groups` will select the same dimension.

Does having random projection dimension equals <= 1 makes sense for this architecture?
It seems like guarding against the univariate case using something like `self.use_rp = False` would be a better choice here. Similar to what you suggested in your pytorch implementation.

**Review by RecreationalMath — CHANGES_REQUESTED** [review]
This is a partial bugfix, please check the comments I left to make it complete.

**Comment by srupat** [comment]
**Summary of changes made**

- Updated the shared TensorFlow TapNet network in `sktime/networks/tapnet.py`.
- Updated the TensorFlow classifier wrapper in `sktime/classification/deep_learning/tapnet.py`.
- Updated the TensorFlow regressor wrapper in `sktime/regression/deep_learning/tapnet.py`.

**API alignment**

- Replaced `rp_params` with `rp_group` and `rp_alpha` on the TensorFlow side to match the current Torch implementation.
- Added `fc_dropout` to the TensorFlow network/classifier/regressor to match the Torch-side output-head configuration.
- Renamed the shared-network hidden activation argument from `activation` to `activation_hidden` for consistency with the Torch implementation and the estimator wrappers.
- Updated the TensorFlow classifier/regressor docstrings to reflect the new parameters and callable activation support.

**Shared network changes**

- Added validation for:
  - `kernel_size`
  - `filter_sizes`
  - `layers`
  - `rp_group`
  - `rp_alpha`
  - the requirement that at least one of `use_lstm` / `use_cnn` is enabled

- Changed random projection sizing to:
  - `rp_dim = floor(n_dims * rp_alpha / rp_group)`
- If that produces `rp_dim < 1`, random projection is disabled with a warning.

Reason:
This matches the newer Torch logic and avoids silently constructing an invalid RP configuration, especially for low-dimensional inputs.

- Switched RP index generation to a deterministic RNG driven by `random_state`.

Reason:
This brings TensorFlow behavior closer to Torch and improves reproducibility.

**CNN path alignment**

- Reworked the TensorFlow CNN branch to build convolution blocks from `kernel_size` / `filter_sizes` generically instead of assuming exactly three fixed conv layers.
- Added actual CNN dropout application in the conv blocks.
- Kept the RP branch structure but updated it to use the same generalized conv-stack approach.

Reason:
The Torch implementation now supports arbitrary conv-stack lengths from the provided tuples, and `dropout` is intended to affect the convolutional path. The previous TensorFlow implementation still had the older fixed-shape logic and did not actually apply CNN dropout.

**Attention alignment**

- Updated LSTM attention to use the LSTM hidden size rather than a hardcoded value.
- Updated CNN attention to use `filter_sizes[-1]` rather than a hardcoded value.
- Kept the same multiplicative attention pattern where applicable.

Reason:
This makes the TensorFlow attention dimensions track the actual network configuration instead of relying on fixed constants.

**Mapping head alignment**

- Updated the TensorFlow mapping section to support arbitrary `layers` lengths instead of assuming exactly two dense layers.
- Added `fc_dropout` before the final estimator output layer.

Reason:
This matches the Torch-side mapping head more closely and makes the TensorFlow backend honor the same configurable dense-head structure.

File: sktime/networks/tapnet.py (hunk 1 above)
Comment (srupat) [reply sktime/networks/tapnet.py]: LMK your thoughts on https://github.com/sktime/sktime/pull/9447#issuecomment-4060660624

File: sktime/networks/tapnet.py (hunk 2 above)
Comment (srupat) [reply sktime/networks/tapnet.py]: Fixed in c8da455

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

Can you please comment why you are:
* removing details from docstrings
* changing some parameters like `rp_...`, and replacing them?

**Comment by srupat** [comment]
> * changing some parameters like `rp_...`, and replacing them?

The change from using `rp_params` to `rp_group` and `rp_alpha` was done for the following reasons:
- `rp_params` was intended to be a 2-int tuple that represented `(number of groups, group dimension)` . However, I realized that this was pretty confusing (for example, setting the first element to -1 will involve calculating the rp_dim) and the calculation involves a hardcoded 2 / 3 factor.
https://github.com/sktime/sktime/blob/9797e604c1ed20e0c4df2b0bff8d8e032d915147/sktime/networks/tapnet.py#L175
- The [original TapNet paper](https://xuczhang.github.io/papers/aaai20_tapnet_full.pdf) mentions using a parameter `alpha` that controls the calculation of `rp_dim`.
image

These changes were also implemented while working on the PyTorch version of the same estimator.
Refer to [comment (1)](https://github.com/sktime/sktime/pull/9412#discussion_r2864629951) and [comment (2)](https://github.com/sktime/sktime/pull/9412#discussion_r2865088543)

**Comment by srupat** [comment]
> * removing details from docstrings

Added details to the docstrings in e0b725f and re-ordered the docstrings in the base, classifier and regressor classes to match the `init` order according to [this](https://github.com/sktime/sktime/pull/9412#discussion_r2831602581) suggestion.

**Comment by srupat** [comment]
@fkiraly Also fixed the persistence failures found by `check_estimator`. 

Changes made:
- `sktime/networks/tapnet/_tapnet_tf.py`: replaced the random-projection `keras.layers.Lambda` with a serializable frozen `Dense` selector, because the Lambda captured TensorFlow state and failed during model persistence.
- `sktime/classification/deep_learning/base/_base_tf.py`: added a `get_custom_objects()` hook and passed it to `load_model`, matching the regressor-side pattern so classifiers with custom Keras layers can be reloaded.
- `sktime/regression/deep_learning/base/_base_tf.py`: passed the existing `get_custom_objects()` hook in the file-based load path too, so disk serialization works consistently with in-memory serialization.
- `sktime/classification/deep_learning/tapnet/_tapnet_tf.py`: added the TapNet classifier custom object mapping for `SeqSelfAttention`.
- `sktime/regression/deep_learning/tapnet/_tapnet_tf.py`: added the TapNet regressor custom object mapping for `SeqSelfAttention`.

I ran `check_estimator` for both `TapNetClassifier` and `TapNetRegressor`, and all tests pass locally.

**Review by fkiraly — CHANGES_REQUESTED** [review]
This changes both the extension contract for the base class and the construction contract. This is not a "resolving minor bugs" PR.

Please review, in case an AI has decided it needs to change the base class.

**Comment by srupat** [comment]
Hi @fkiraly 
The scope for this PR has definitely extended as compared to what it was intended for earlier. 

However, there are a few things
- We need a `get_custom_objects` for the base classifier class. This is because we need to load the custom Keras layer `SeqSelfAttention` for TapNet. Hence, we need to pass it as `custom_objects` like so
```python
keras.models.load_model(path, custom_objects={"SeqSelfAttention": SeqSelfAttention})
```
The concrete child classes (like `TapNetClassifier` then override this method and just pass on to the custom layer like this `SeqSelfAttention.get_custom_objects()`
- This has already been implemented in the base Regressor class, and was something that is missing from the classifier class.
https://github.com/sktime/sktime/blob/5a7adfdfe50c7cad8837c7e2ef1679458dab169a/sktime/regression/deep_learning/base/_base_tf.py#L246-L252

I am thinking of adding this as a seperate issue, rather than it being a part of this PR. After that is merged, this PR can then be merged succesfully.

**Comment by srupat** [comment]
@fkiraly I noticed that these classifiers skip all tests by setting this tag
`"tests:skip_all": True,`. Does this need to stay? These estimators can maybe be included in the tests after bugs like #10537 are fixed.

**Review by fkiraly — APPROVED** [review]
Great!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Wait a second, does this depend on https://github.com/sktime/sktime/pull/10637?

**Comment by srupat** [comment]
@fkiraly 
Yes, this depends on #10637 for `test_persistence_via_pickle`/`test_save_estimators_to_file` tests to fully pass (both use the `SeqSelfAttention` custom layer, which needs the `_get_keras_custom_objects` base-class wiring added there). 

I've written `TapNetClassifier`/`TapNetRegressor`'s overrides to already match that method name, and verified locally (by temporarily simulating #10637's base-class change) that everything passes once it's in place — currently only those two test groups fail, for exactly that reason, nothing else. Rather than opening a second PR, I'll pull #10637's changes into this branch once it merges to `main`.

**Comment by srupat** [comment]
Fixes #3539

**Comment by srupat** [comment]
About 3284642:
- Removed `tests:skip_all` and failing tests from tags. Only one test remains, however that would be fixed once #10453 is merged. Check #10712 
- Renamed `get_custom_objects` to `_get_keras_custom_objects` to support #10637 once it's merged.
- Added missing example docstring to TapNet regressor
- Removed TapNetRegressor and failing tests from legacy `sktime/tests/_config.py`

**Comment by srupat** [comment]
@fkiraly , merged with main, all tests pass locally. 

One question -- should the `tests:vm = True` tag be removed from `TapNetClassifier`? I think this is related to #8518 , but is it due to some other tests (core tests) failing or due to architectural failures?

**Comment by fkiraly** [comment]
> One question -- should the `tests:vm = True` tag be removed from `TapNetClassifier`?

No, this tag just lets the estimator tests run on their own VM with specific dependencies installed.

**Comment by srupat** [comment]
@fkiraly added doctest example to the network class `TapNetNetwork`  in 854d61f. I had locally run the test suite for the classifier and regressor, but not the network class. Thanks for tracking this closely and apologies for the oversight.

**Review by fkiraly — APPROVED** [review]
Perfect, thanks!
